File: frontend/src/api/library.ts
import { api } from './client'

export interface LibraryTrack {
  id: string
  title: string
  artist: string
  album: string | null
  year: number | null
  cover_art_url: string | null
  file_path: string | null
  file_size: number | null
  duration_ms: number | null
}

export interface AlbumInfo {
  album: string
  artist: string
  cover_art_url: string | null
  year: number | null
  track_count: number
}

export interface ArtistInfo {
  artist: string
  track_count: number
}

export async function fetchTracks(params?: {
  search?: string
  artist?: string
  album?: string
  page?: number
  limit?: number
}): Promise<LibraryTrack[]> {
  const resp = await api.get<LibraryTrack[]>('/library/tracks', { params })
  return resp.data
}

export async function fetchAlbums(search?: string): Promise<AlbumInfo[]> {
  const resp = await api.get<AlbumInfo[]>('/library/albums', { params: { search } })
  return resp.data
}

export async function fetchArtists(): Promise<ArtistInfo[]> {
  const resp = await api.get<ArtistInfo[]>('/library/artists')
  return resp.data
}

export function getStreamUrl(trackId: string): string {
  return `/api/files/${trackId}/stream`
}

File: backend/app/api/library.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func, distinct
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import Track
from app.schemas import LibraryTrack

router = APIRouter()


@router.get("/tracks", response_model=list[LibraryTrack])
async def list_tracks(
    search: str = Query(default="", description="Search by title, artist, or album"),
    artist: str = Query(default="", description="Filter by artist"),
    album: str = Query(default="", description="Filter by album"),
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
):
    query = select(Track).where(Track.status == "completed", Track.file_path.isnot(None))

    if search:
        pattern = f"%{search}%"
        query = query.where(
            (Track.title.ilike(pattern)) |
            (Track.artist.ilike(pattern)) |
            (Track.album.ilike(pattern))
        )
    if artist:
        query = query.where(Track.artist.ilike(f"%{artist}%"))
    if album:
        query = query.where(Track.album.ilike(f"%{album}%"))

    query = query.order_by(Track.created_at.desc()).offset((page - 1) * limit).limit(limit)
    result = await db.execute(query)
    return result.scalars().all()


@router.get("/albums")
async def list_albums(
    search: str = Query(default=""),
    db: AsyncSession = Depends(get_db),
):
    query = select(
        Track.album,
        Track.artist,
        Track.cover_art_url,
        Track.year,
        func.count(Track.id).label("track_count"),
    ).where(
        Track.status == "completed",
        Track.file_path.isnot(None),
        Track.album.isnot(None),
        Track.album != "",
    ).group_by(Track.album, Track.artist)

    if search:
        query = query.where(Track.album.ilike(f"%{search}%"))

    query = query.order_by(Track.album)
    result = await db.execute(query)
    rows = result.all()
    return [
        {
            "album": r.album,
            "artist": r.artist,
            "cover_art_url": r.cover_art_url,
            "year": r.year,
            "track_count": r.track_count,
        }
        for r in rows
    ]


@router.get("/artists")
async def list_artists(db: AsyncSession = Depends(get_db)):
    query = select(
        Track.artist,
        func.count(Track.id).label("track_count"),
    ).where(
        Track.status == "completed",
        Track.file_path.isnot(None),
    ).group_by(Track.artist).order_by(Track.artist)

    result = await db.execute(query)
    rows = result.all()
    return [{"artist": r.artist, "track_count": r.track_count} for r in rows]
